spec/specification.md

```diff
@@ -0,0 +1,639 @@
+
+e += (Frame typ e) # a special expression for assisting with evaluation
+```
+
+### Static Semantics
+
+#### Notation and Concepts
+
+Typing judgments are of the form:
+```
+C |- expr : typ
+```
+which is read as: "within context C, it holds that `expr` is of type `typ`".
+
+The typing rules are given as *deduction rules* of the following form:
+```
+premise_1   premise_2
+---------------------
+     conclusion
+```
+which is read as: "if `premise_1` and `premise_2` hold true, it follows that
+`conclusion` also holds true".
+
+A deduction rule without a premise is called an *axiom*; the conclusion always
+holds true.
+
+#### Type Relations
+
+All type equality uses *structural equality*. The order of types in a `TupleTy`
+and `ProcTy` is significant, in `UnionTy` it is not. Formally:
+```
+void  = void
+unit  = unit
+bool  = bool
+int   = int
+float = float
+
+(TupleTy a_0 ... a_n) = (TupleTy b_0 ... b_n)  (where a_0 = b_0 ... a_n = b_n ^ n >= 1)
+(ProcTy  a_0 ... a_n) = (ProcTy  b_0 ... b_n)  (where a_0 = b_0 ... a_n = b_n)
+(UnionTy a_0 ... a_n) = (UnionTy b_0 ... b_n)  (where {a_0 ... a_n} = {b_0 ... b_0})
+```
+
+`a <: b` means "a is a subtype of b". `a <:= b` means "a is a subtype of or
+equal to b".
+```
+# void is the bottom type; it's a subtype of all other types
+void <: unit
+void <: bool
+void <: int
+void <: float
+void <: (TupleTy typ+)
+void <: (UnionTy typ+)
+void <: (ProcTy typ+)
+
+typ_1 <: (UnionTy typ_1 typ_2 ...) # | a type is a subtype of `UnionTy` if it's
+                                   # | a part of the union (in any position)
+```
+
+#### Typing Rules
+
+This section describes how types are assigned to the abstract syntax. An
+abstract syntax tree is well-formed if and only if a type can be assigned to
+it.
+
+`C` (the context) is a record with the following abstract syntax:
+```
+C ::= { symbols S
+        return typ }
+```
+
+`S` is the function mapping identifiers to types. `return` is the return type
+of the procedure being judged.
+
+Unless explicitly stated otherwise, a deduction rule's conclusion is always
+a single *typing judgement*, whereas a premise may be either a
+*typing judgment* or *side condition*.
+
+Type expressions use a separate set of rules. They're identified by the
+conclusion being a judgments of the form `C |-_t x : typ`.
+
+##### Primitive Types
+
+```
+
+--------------------- # S-void-type
+C |-_t (VoidTy) : void
+
+
+--------------------- # S-unit-type
+C |-_t (UnitTy) : unit
+
+
+--------------------- # S-bool-type
+C |-_t (BoolTy) : bool
+
+
+------------------- # S-int-type
+C |-_t (IntTy) : int
+
+
+----------------------- # S-float-type
+C |-_t (FloatTy) : float
+
+
+----------------------- # S-empty-tuple-type
+C |-_t (TupleTy) : unit
+```
+
+##### Composite Types
+
+```
+x in C.symbols  C.symbols(x) = (type typ)
+----------------------------------------- # S-type-ident
+C |-_t x : typ
+
+C |-_t e : typ ...  typ != void ...
```

Comment:
Within deduction rules, I used them to mean "repeat judgment, side-condition, or whatever else they're trailing for every item the input pattern matched".

Consider the following made-up rule:
```
C |- e : int ...
---------------------
C |- (Node e+) : node
```
`e` refers to one or more expressions here (because of the `e+` in the conclusion's pattern), and the ellipsis following the judgment in the premise is meant to highlight that the judgment is repeated (if `e` matched more than one element, that is). In a way, this makes the rule somewhat of a rule template.

Ultimately, using a macro DSL is going to resolve these problems with ad-hoc, but if you have a suggestion for how to better express the above in the meantime, I'd be happy to hear.